File: test-zkevm/hardhat2/scripts/hello.ts
import { ethers } from "hardhat";
import { HardhatRuntimeEnvironment } from "hardhat/types";
import { Deployer } from "@matterlabs/hardhat-zksync-deploy";
import { Provider as ZkProvider, Wallet as ZkWallet } from "zksync-ethers";
import * as dotenv from "dotenv";

dotenv.config();

function getPrivateKeyFromMnemonic(mnemonic: string, index: number = 0): { privateKey: string, publicKey: string, address: string } {
  // Create the HDNode from the mnemonic
  const hdNode = ethers.HDNodeWallet.fromPhrase(mnemonic);

  // Derive the account directly using the full path
  const fullPath = `m/44'/60'/0'/0/${index}`;
  // ethers memnoics from mnemonics
  let ethersMnemonic = ethers.Mnemonic.fromPhrase(mnemonic);
  const wallet = ethers.HDNodeWallet.fromMnemonic(ethersMnemonic, fullPath);

  return {
      privateKey: wallet.privateKey.slice(2),
      publicKey: wallet.publicKey.slice(2),
      address: wallet.address.slice(2)
  };
}

async function create_contract(hre: HardhatRuntimeEnvironment) {
  console.log(`Running deploy script for the Hello contract`);

  // Get the network configuration
  const network = hre.network.config as any;

  // Initialize the providers
  const l1Provider = new ethers.JsonRpcProvider(network.ethNetwork);
  const l2Provider = new ZkProvider(network.url);

  // Get the mnemonic from the network configuration
  const mnemonic = network.accounts.mnemonic;

  // Use the getPrivateKeyFromMnemonic function to get the private key
  const { privateKey } = getPrivateKeyFromMnemonic(mnemonic);

  const zkWallet = new ZkWallet(privateKey, l2Provider, l1Provider);

  // Create deployer object
  const deployer = new Deployer(hre, zkWallet);

  // Load the artifact of the contract you want to deploy.
  const artifact = await deployer.loadArtifact("HelloWorld");

  // Estimate contract deployment fee
  const deploymentFee = await deployer.estimateDeployFee(artifact, []);

  // Deploy this contract. The returned object will be of a `Contract` type, similarly to ones in `ethers`.
  const parsedFee = ethers.formatEther(deploymentFee.toString());
  console.log(`The deployment is estimated to cost ${parsedFee} ETH`);

  const helloContract = await deployer.deploy(artifact, []);

  // Show the contract info.
  const contractAddress = await helloContract.getAddress();
  console.log(`${artifact.contractName} was deployed to ${contractAddress}`);

  // Assuming you have the contract ABI and provider
  const contractABI = artifact.abi;
  const provider = l2Provider;
  const helloWorldContract = new ethers.Contract(contractAddress, contractABI, zkWallet);
  
  // Get the initial greeting
  let greeting = await helloWorldContract.getGreeting();
  console.log("Initial Greeting:", greeting);

  // Set a new greeting
  const newGreeting = "Hello from zkSync!";
  await helloWorldContract.setGreeting(newGreeting);
  console.log("Setting new greeting...");

  // Get the updated greeting
  greeting = await helloWorldContract.getGreeting();
  console.log("Updated Greeting:", greeting);
}

async function main() {
  console.log("Hello, World!");

  // Get 5 accounts from the Hardhat network
  const accounts = await hre.ethers.getSigners();

  for (let i = 0; i < 5; i++) {
    const account = accounts[i];
    console.log(`Account ${i + 1}:`);
    console.log(account);
    console.log();
  }

  // Call create_contract function with hre
  await create_contract(hre);
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: test-zkevm/hardhat2/scripts/hello.py
import os
import json
from dotenv import load_dotenv
from web3 import Web3
from eth_account import Account
from eth_account.signers.local import LocalAccount
from zksync2.module.module_builder import ZkSyncBuilder
from zksync2.signer.eth_signer import PrivateKeyEthSigner
from zksync2.transaction.transaction_builders import TxCreateContract
from zksync2.manage_contracts.contract_encoder_base import ContractEncoder
from zksync2.core.types import EthBlockParams
import binascii
from eth_account.signers.local import LocalAccount

load_dotenv()

async def deploy_contract(zk_web3, account: LocalAccount, abi, bytecode):
    print("Running deploy script for the Hello contract")

    # Convert bytecode to bytes if it's a string
    #if isinstance(bytecode, str):
    #    bytecode = bytes.fromhex(bytecode.strip('0x'))
    if isinstance(bytecode, str):
        bytecode = bytes.fromhex(bytecode)

    # Pad the bytecode if necessary
    if len(bytecode) % 64 == 0:
        bytecode += b'\0'  # Add a single zero byte

    # Get chain id of zkSync network
    chain_id = zk_web3.zksync.chain_id

    # Signer is used to generate signature of provided transaction
    signer = PrivateKeyEthSigner(account, chain_id)

    # Get nonce of ETH address on zkSync network
    nonce = zk_web3.zksync.get_transaction_count(
        account.address, EthBlockParams.PENDING.value
    )

    # Get current gas price in Wei
    gas_price = zk_web3.zksync.gas_price

    
    # Create deployment contract transaction
    create_contract = TxCreateContract(
        web3=zk_web3,
        chain_id=chain_id,
        nonce=nonce,
        from_=account.address,
        gas_limit=0,  # UNKNOWN AT THIS STATE
        gas_price=gas_price,
        bytecode=bytecode,
    )
    
    

    # ZkSync transaction gas estimation
    estimate_gas = zk_web3.zksync.eth_estimate_gas(create_contract.tx)
    print(f"Fee for transaction is: {Web3.from_wei(estimate_gas * gas_price, 'ether')} ETH")

    # Convert transaction to EIP-712 format
    tx_712 = create_contract.tx712(estimate_gas)

    # Sign message
    signed_message = signer.sign_typed_data(tx_712.to_eip712_struct())

    # Encode signed message
    msg = tx_712.encode(signed_message)

    # Deploy contract
    tx_hash = zk_web3.zksync.send_raw_transaction(msg)

    # Wait for deployment contract transaction to be included in a block
    tx_receipt = zk_web3.zksync.wait_for_transaction_receipt(
        tx_hash, timeout=240, poll_latency=0.5
    )

    print(f"Tx status: {tx_receipt['status']}")
    contract_address = tx_receipt["contractAddress"]

    print(f"Deployed contract address: {contract_address}")

    return contract_address

Account.enable_unaudited_hdwallet_features()
async def main():
    print("Hello, World!")
    mnemonic = os.getenv('MYMNEMONICS')
    if not mnemonic:
        return {"status": "error", "message": "MYMNEMONICS environment variable not set"}
    
    current_wallet_index = 0
    from_account: LocalAccount = Account.from_mnemonic(mnemonic, account_path=f"m/44'/60'/0'/0/{current_wallet_index}")
    
    private_key = from_account._private_key.hex()
    address = from_account.address
    
    print(f"Derived address: {address}")
    
    # Get network configuration
    network_config = {
        'url': 'https://testnet.zkevm.cronos.org',
        'from': address,
        'private_key': private_key
    }

    # Load the artifact of the contract you want to deploy
    with open('artifacts-zk/contracts/HelloWorld.sol/HelloWorld.json', 'r') as file:
        contract_json = json.load(file)

    abi = contract_json['abi']
    bytecode = contract_json['bytecode']
    if bytecode.startswith('0x'):
        bytecode = bytecode[2:]

    # Connect to zkSync network
    zk_web3 = ZkSyncBuilder.build(network_config['url'])

    # Get account object by providing from private key
    account: LocalAccount = Account.from_key(private_key)

    # Deploy the contract
    deployed_address = await deploy_contract(zk_web3, account, abi, bytecode)
    print(f"Contract deployed at: {deployed_address}")

    # Create contract instance
    contract = zk_web3.eth.contract(address=deployed_address, abi=abi)

    # Get initial greeting
    initial_greeting = contract.functions.getGreeting().call()
    print(f"Initial greeting: {initial_greeting}")

    # Set new greeting
    new_greeting = "Hello from zkSync!"
    tx = contract.functions.setGreeting(new_greeting).build_transaction({
        'from': address,
        'nonce': zk_web3.eth.get_transaction_count(address),
    })
    signed_tx = zk_web3.eth.account.sign_transaction(tx, private_key)
    tx_hash = zk_web3.eth.send_raw_transaction(signed_tx.rawTransaction)
    tx_receipt = zk_web3.eth.wait_for_transaction_receipt(tx_hash)
    print(f"New greeting set: {new_greeting}")

    # Get updated greeting
    updated_greeting = contract.functions.getGreeting().call()
    print(f"Updated greeting: {updated_greeting}")

if __name__ == "__main__":
    import asyncio
    asyncio.run(main())
